php_phongo.c

```diff
@@ -2841,18 +2841,379 @@ static bool phongo_manager_set_auto_encryption_opts(php_phongo_manager_t* manage
+			phongo_throw_exception(PHONGO_ERROR_INVALID_ARGUMENT TSRMLS_CC, "Expected keyAltNames to be array, %s given", PHONGO_ZVAL_CLASS_OR_TYPE_NAME_P(zkeyaltnames));
+			goto cleanup;
+		}
+
+		ht_data           = HASH_OF(zkeyaltnames);
+		keyaltnames_count = ht_data ? zend_hash_num_elements(ht_data) : 0;
+		keyaltnames       = bson_malloc(sizeof(char*) * keyaltnames_count);
+
+#if PHP_VERSION_ID >= 70000
+		{
+			zval* keyaltname;
+
+			ZEND_HASH_FOREACH_VAL(ht_data, keyaltname)
+			{
+				if (i == keyaltnames_count) {
+					phongo_throw_exception(PHONGO_ERROR_LOGIC TSRMLS_CC, "Iterating over too many keyAltNames. Please file a bug report");
+					failed = true;
+				}
+
+				keyaltnames[i] = bson_strdup(Z_STRVAL_P(keyaltname));
+				i++;
+			}
+			ZEND_HASH_FOREACH_END();
+		}
+#else
+		{
+			HashPosition pos;
+			zval**       keyaltname;
+			bool         failed = false;
+
+			for (
+				zend_hash_internal_pointer_reset_ex(ht_data, &pos);
+				zend_hash_get_current_data_ex(ht_data, (void**) &keyaltname, &pos) == SUCCESS;
+				zend_hash_move_forward_ex(ht_data, &pos)) {
+
+				if (i == keyaltnames_count) {
+					phongo_throw_exception(PHONGO_ERROR_LOGIC TSRMLS_CC, "Iterating over too many keyAltNames. Please file a bug report");
+					failed = true;
+					break;
+				}
+
+				keyaltnames[i] = bson_strdup(Z_STRVAL_PP(keyaltname));
+				i++;
+			}
+		}
+#endif
+
+		if (!failed) {
+			mongoc_client_encryption_datakey_opts_set_keyaltnames(opts, keyaltnames, keyaltnames_count);
+		}
+
+		for (i = 0; i < keyaltnames_count; i++) {
+			bson_free(keyaltnames[i]);
+		}
+		bson_free(keyaltnames);
+
+		if (failed) {
+			goto cleanup;
+		}
+	}
+
+	if (php_array_existsc(options, "masterKey")) {
+		bson_t masterkey = BSON_INITIALIZER;
+
+		php_phongo_zval_to_bson(php_array_fetchc(options, "masterKey"), PHONGO_BSON_NONE, &masterkey, NULL TSRMLS_CC);
+		if (EG(exception)) {
+			goto cleanup;
+		}
+
+		mongoc_client_encryption_datakey_opts_set_masterkey(opts, &masterkey);
+	}
+
+	return opts;
+
+cleanup:
+	if (opts) {
+		mongoc_client_encryption_datakey_opts_destroy(opts);
+	}
+
+	return NULL;
+} /* }}} */
+
+void phongo_clientencryption_create_datakey(php_phongo_clientencryption_t* clientencryption, zval* return_value, char* kms_provider, zval* options TSRMLS_DC) /* {{{ */
+{
+	mongoc_client_encryption_datakey_opts_t* opts;
+	bson_value_t                             keyid;
+	bson_error_t                             error = { 0 };
+
+	if (!(opts = phongo_clientencryption_datakey_opts_from_zval(options TSRMLS_CC))) {
+		/* Exception already thrown */
+		goto cleanup;
+	}
+
+	if (!mongoc_client_encryption_create_datakey(clientencryption->client_encryption, kms_provider, opts, &keyid, &error)) {
+		phongo_throw_exception_from_bson_error_t(&error TSRMLS_CC);
+		goto cleanup;
+	}
+
+	if (!php_phongo_bson_value_to_zval(&keyid, return_value)) {
+		phongo_throw_exception(PHONGO_ERROR_UNEXPECTED_VALUE TSRMLS_CC, "Could not convert BSON value to ZVAL");
```

Comment:
Is this exception necessary? Looking at `php_phongo_bson_value_to_zval`, it's only error condition is when `php_phongo_bson_to_zval_ex` itself fails and in that case an exception will have been thrown. This code would end up stacking another exception on top of that and may hide the more relevant error message.

Reply:
Good point. I've also noted a wrong return statement in the `cleanup` block of `php_phongo_bson_value_to_zval` and fixed that.